VariantEncoding.md

```diff
@@ -0,0 +1,429 @@
+| Exact Numeric        | decimal16                   | `10`    | DECIMAL(precision, scale)   | 1 byte scale in range [0, 38], followed by little-endian unscaled value (see decimal table)                         |
+| Date                 | date                        | `11`    | DATE                        | 4 byte little-endian                                                                                                |
+| Timestamp            | timestamp                   | `12`    | TIMESTAMP(true, MICROS)     | 8-byte little-endian                                                                                                |
+| TimestampNTZ         | timestamp without time zone | `13`    | TIMESTAMP(false, MICROS)    | 8-byte little-endian                                                                                                |
+| Float                | float                       | `14`    | FLOAT                       | IEEE little-endian                                                                                                  |
+| Binary               | binary                      | `15`    | BINARY                      | 4 byte little-endian size, followed by bytes                                                                        |
+| String               | string                      | `16`    | STRING                      | 4 byte little-endian size, followed by UTF-8 encoded bytes                                                          |
+
+| Decimal Precision     | Decimal value type |
+|-----------------------|--------------------|
+| 1 <= precision <= 9   | int32              |
+| 10 <= precision <= 18 | int64              |
+| 18 <= precision <= 38 | int128             |
+| > 38                  | Not supported      |
+
+The *Logical Type* column indicates logical equivalence of physically encoded types.
+For example, a user expression operating on a string value containing "hello" should behave the same, whether it is encoded with the short string optimization, or long string encoding.
+Similarly, user expressions operating on an *int8* value of 1 should behave the same as a decimal16 with scale 2 and unscaled value 100.
+
+# Field ID order and uniqueness
+
+For objects, field IDs and offsets must be listed in the order of the corresponding field names, sorted lexicographically.
+Note that the fields themselves are not required to follow this order.
+As a result, offsets will not necessarily be listed in ascending order.
+
+An implementation may rely on this field ID order in searching for field names.
+E.g. a binary search on field IDs (combined with metadata lookups) may be used to find a field with a given field.
+
+Field names are case-sensitive.
+Field names are required to be unique for each object.
+It is an error for an object to contain two fields with the same name, whether or not they have distinct dictionary IDs.
```

Comment:
modulo case differences (I know this is stated above that names are case sensitive but i just want to make sure I am parsing this correctly)?